msbuild/Xamarin.MacDev.Tasks/Tasks/CalculateAssembliesReport.cs

```diff
@@ -0,0 +1,65 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+using System.Text;
+using Microsoft.Build.Framework;
+using Xamarin.Messaging.Build.Client;
+
+namespace Xamarin.MacDev.Tasks {
+	public class CalculateAssembliesReport : XamarinTask {
+		[Required]
+		public string WorkingDirectory { get; set; } = string.Empty;
+
+		[Required]
+		public string TargetReportFile { get; set; } = string.Empty;
+
+		public override bool Execute ()
+		{
+			if (ShouldExecuteRemotely ()) {
+				return new TaskRunner (SessionId, BuildEngine4).RunAsync (this).Result;
+			}
+
+			if (!Directory.Exists (WorkingDirectory)) {
+				Log.LogError ("Unable to calculate the assemblies report from '{0}'. Directory not found.", WorkingDirectory);
+
+				return false;
+			}
+
+			try {
+				var reportEntriesBuilder = new StringBuilder ();
+				IEnumerable<string> files = Directory
+					.GetFiles (WorkingDirectory, "*", SearchOption.TopDirectoryOnly)
+					.Where (f => Path.GetExtension (f) == ".dll");
+
+				foreach (var file in files) {
+					try {
+						var fileInfo = new FileInfo (file);
+						using Stream stream = fileInfo.OpenRead ();
+						using var peReader = new PEReader (stream);
+						MetadataReader metadataReader = peReader.GetMetadataReader ();
+						Guid mvid = metadataReader.GetGuid (metadataReader.GetModuleDefinition ().Mvid);
```

Comment:
Just out of curiosity, how long does this task take computing the MVID for all these assemblies (as opposed to computing the hash)?

Reply:
I would say around 40% faster when creating the file with MVID calculation, based on my local binlogs. It still takes time though, we might want consider assemblies filtering as a potential perf improvement as a next step